horovod/spark/keras/datamodule.py

```diff
@@ -0,0 +1,166 @@
+                                transform_spec=self.transform_spec,
+                                storage_options=self.storage_options,
+                                shuffle_row_groups=False,
+                                **self.reader_factory_kwargs) if self.has_val else self.empty_batch_reader()
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if self.has_val and self.val_reader:
+            self.val_reader.stop()
+        if self.train_reader:
+            self.train_reader.stop()
+        super().__exit__(type, value, traceback)
+
+    @contextlib.contextmanager
+    def empty_batch_reader(self):
+        yield None
+
+    def train_data(self):
+        return self.make_dataset(self.train_reader, self.train_batch_size, self.shuffle_size,
+                                 self.is_batch_reader, shuffle=True if self.shuffle_size > 0 else False,
+                                 cache=self.inmemory_cache_all, seed=self.random_seed)
+
+    def val_data(self):
+        return self.make_dataset(self.val_reader, self.val_batch_size, self.shuffle_size,
+                                 self.is_batch_reader, shuffle=False, cache=self.inmemory_cache_all) if self.val_reader else None
+
+
+class NVTabularDataModule(DataModule):
+    """NVTabular-based DataModule for KerasEstimator for GPU-accelerated data loading of tabular datasets.
+
+    Note: requires `label_cols`, `categorical_cols`, and `continuous_cols` to be explicitly provided."""
+    def __init__(self, label_cols=[], categorical_cols=[], continuous_cols=[], **kwargs):
+        super().__init__(**kwargs)
+        self.label_cols = label_cols
+        self.categorical_cols = categorical_cols
+        self.continuous_cols = continuous_cols
+
+    @staticmethod
+    def seed_fn():
+        """
+        Generate consistent dataloader shuffle seeds across workers
+        Reseeds each worker's dataloader each epoch to get fresh a shuffle
+        that's consistent across workers.
+        """
+        import horovod.tensorflow.keras as hvd
+        import numpy as np
+        import tensorflow as tf
+        hvd.init()
+        seed = np.random.randint(0, tf.int32.max)
+        seed_tensor = tf.constant(seed)
+        root_seed = hvd.broadcast(seed_tensor, name="shuffle_seed", root_rank=0)
+        return root_seed
+
+    @staticmethod
+    def to_dense(X, labels):
+        """Convert NVTabular's ragged representation of array types to dense arrays."""
+        import tensorflow as tf
+        retX = {}
+        for col, tensor in X.items():
+            if isinstance(tensor, tuple):
+                retX[col] = tf.RaggedTensor.from_row_lengths(values=tensor[0][:,0], row_lengths=tensor[1][:,0]).to_tensor()
+            else:
+                retX[col] = tensor
+        return retX, labels
+
+    def train_data(self):
+        import horovod.tensorflow.keras as hvd
+        from nvtabular.loader.tensorflow import KerasSequenceLoader
```

Comment:
Does this line require `nvtabular` to be a required package of horovod?
In that case, we should change `setup.py` part https://github.com/horovod/horovod/blob/master/setup.py#L169 ?

Reply:
As long as `dm.train_data()` with NVTabular data module is not called, we are safe then?

Reply:
Yes, I would say using `NVTabularDataModule` requires `nvtabular` to be installed, but I would not make it a dependency. If you are not calling into this code path, `nvtabular` is not required.